### PR TITLE
btrfs-progs: mkfs/rootdir: add hard link support

### PR DESCRIPTION
**Pull request by adam900710**
With the recently reworked --rootdir support, although it solves several
hard link related problems, it splits the hard links into new inodes.

And on each split, it shows a warning on each file with hardlinks.

Although the split behavior doesn't cause any data corruption, it can
still be pretty noisy for rootfs creation, as there are a lot of distros
storing timezone files as hardlinks.

This patchset adds back the hard link detection and creation, with
enhanced handling to co-operate with --subvol option.

The details can be found in the first patch, with the new corner case
introduced by --subvol option.

The second patch enhances the existing --rootdir and --subvol test case
with extra corner cases like hard links, and hard links split by
subvolume boundary.

### CODE REVIEW

**Comment by kdave** [comment]
Can you please rebase the branch? The diff is mixed with all the v6.10.1 changes, I can't do the review properly.

**Comment by kdave** [comment]
Never mind, it was caused by my push of old devel, before the 6.10.1 merge.

File: Documentation/mkfs.btrfs.rst [hunk 1]
```diff
@@ -160,6 +160,19 @@ OPTIONS
...
+	If there are hard links inside *rootdir* and *subdir* will split the
+	subvolumes, like the following case::
+
+		rootdir/
+		|- hardlink1
+		|- hardlink2
+		|- subdir/  <- will be a subvolume
+		   |- hardlink3
+
+	In that case we can not create `hardlink3` as a hardlinks of
```
Comment (kdave) [inline Documentation/mkfs.btrfs.rst:172]: can not -> cannot, 'as hardlinks of'

File: Documentation/mkfs.btrfs.rst [hunk 2]
```diff
@@ -160,6 +160,19 @@ OPTIONS
...
+	`hardlink1` and `hardlink2`.
+	Because hardlink3 will be inside a new subvolume.
```
Comment (kdave) [inline Documentation/mkfs.btrfs.rst:174]: you can join the sentence, it's logically continuing the previous one

File: mkfs/rootdir.c [hunk 3]
```diff
@@ -74,6 +75,50 @@ struct inode_entry {
...
+ * rootdir/
+ * |- foobar_hardlink1
+ * |- foobar_hardlink2
+ * |- subv/	<- Will be created as a subvolume
+ *    |- foobar_hardlink3.
+ *
+ * Since all the 3 hard links are inside the same rootdir and the same
+ * filesystem, on the host fs they are all hard links to the same inode.
+ *
+ * But for created btrfs, only hardlink1 and hardlink2 can be created as
```
Comment (kdave) [inline mkfs/rootdir.c:95]: Please rephrase "for created btrfs", I'm not sure I understand what you mean.

File: mkfs/rootdir.c [hunk 4]
```diff
@@ -74,6 +75,50 @@ struct inode_entry {
...
+ * hardlinks. Since we can not create hardlink across subvolume.
+ */
+struct hardlink_entry {
+	struct rb_node node;
+	/*
+	 * The following two are reported from the stat() of the
+	 * host filesystem.
+	 *
+	 * For st_nlink we can not trust it unconditionally, as
```
Comment (kdave) [inline mkfs/rootdir.c:104]: can not -> cannot

File: mkfs/rootdir.c [hunk 5]
```diff
@@ -133,6 +175,82 @@ static int rootdir_path_push(struct rootdir_path *path, struct btrfs_root *root,
...
+
+static struct hardlink_entry *find_hard_link(struct btrfs_root *root,
+					     const struct stat *st)
+{
+	struct rb_node *node;
+	struct hardlink_entry tmp;
+
+	tmp.st_dev = st->st_dev;
+	tmp.st_ino = st->st_ino;
+	tmp.root = root;
```
Comment (kdave) [inline mkfs/rootdir.c:212]: You could initialize tmp at the declaration and make it const.

File: mkfs/rootdir.c [hunk 6]
```diff
@@ -620,6 +719,36 @@ static int ftw_add_inode(const char *full_path, const struct stat *st,
 	parent = rootdir_path_last(&current_path);
 	root = parent->root;
 
+	/* Check if there is already a hard link record for this. */
+	if (have_hard_links) {
+		struct hardlink_entry *found;
+		found = find_hard_link(root, st);
```
Comment (kdave) [inline mkfs/rootdir.c:725]: Please insert a newline between declaration and statement

File: tests/mkfs-tests/036-rootdir-subvol/test.sh [hunk 7]
```diff
@@ -11,23 +11,75 @@ prepare_test_dev
...
+	run_check mkdir "$tmp/dir"
+	run_check ln "$tmp/hl1" "$tmp/dir/hl3"
+
+	run_check_mkfs_test_dev --rootdir "$tmp"
+	run_check $SUDO_HELPER "$TOP/btrfs" check "$TEST_DEV"
+
+	run_check_mount_test_dev
+	nr_hardlink=$(run_check_stdout $SUDO_HELPER stat -c "%h" "$TEST_MNT/hl1")
+
+	if [ $nr_hardlink -ne 3 ]; then
```
Comment (kdave) [inline tests/mkfs-tests/036-rootdir-subvol/test.sh:50]: "$nr_hardlink"

**Review by kdave — CHANGES_REQUESTED** [review]

File: mkfs/rootdir.c (hunk 3 above)
Comment (adam900710) [reply mkfs/rootdir.c]: I changed it to "But for the btrfs we are building".

File: tests/mkfs-tests/036-rootdir-subvol/test.sh [hunk 8]
```diff
@@ -11,23 +11,75 @@ prepare_test_dev
...
+	run_check mkdir "$tmp/subv"
+	run_check ln "$tmp/hl1" "$tmp/subv/hl3"
+
+	run_check_mkfs_test_dev --rootdir "$tmp" --subvol subv
+	run_check $SUDO_HELPER "$TOP/btrfs" check "$TEST_DEV"
+
+	run_check_mount_test_dev
+	nr_hardlink=$(run_check_stdout $SUDO_HELPER stat -c "%h" "$TEST_MNT/hl1")
+
+	if [ $nr_hardlink -ne 2 ]; then
```
Comment (kdave) [inline tests/mkfs-tests/036-rootdir-subvol/test.sh:70]: shell variable quoting

File: tests/mkfs-tests/036-rootdir-subvol/test.sh [hunk 9]
```diff
@@ -11,23 +11,75 @@ prepare_test_dev
...
+		_fail "hard link number incorrect for hl1, has ${nr_hardlink} expect 2"
+	fi
+
+	nr_hardlink=$(run_check_stdout $SUDO_HELPER stat -c "%h" "$TEST_MNT/subv/hl3")
+	if [ $nr_hardlink -ne 1 ]; then
```
Comment (kdave) [inline tests/mkfs-tests/036-rootdir-subvol/test.sh:75]: shell variable quoting

**Review by kdave — APPROVED** [review]
Two minor fixes you can do once you merge it. Thanks.
